### DISCUSSION TITLE
What would be a good starting point for writing a code to calculate the gap for a given pressure distribution p=p(x,y)??

### DISCUSSION THREAD

**Discussion by sitangshugk95** [ContactEngineering/ContactMechanics#85] [Q&A]
Thanks to a lot of help from @jotelha and @sannant , I am finally being able to run all packages and codes using the Docker images. 

As a next step, I want to calculate the gap between a rough surface and an elastic flat for a given pressure distribution, p(x,y). N.B. This is different from defining an average pressure, i.e. different discretized grid point will have different pressure values as input. What would be a good starting point for this?

Are there any sample codes written that do something like this (maybe displacement distributions) and can be modified?

**Comment by sannant** [comment]
Ok so this means you do not even need to solve the contact problem.

All you need is the elastic halfspace.

Here are the methods you are looking for:

https://contactengineering.github.io/ContactMechanics/source/ContactMechanics.html?highlight=periodicfft#ContactMechanics.FFTElasticHalfSpace.PeriodicFFTElasticHalfSpace.evaluate_force

https://contactengineering.github.io/ContactMechanics/source/ContactMechanics.html?highlight=periodicfft#ContactMechanics.FFTElasticHalfSpace.PeriodicFFTElasticHalfSpace.evaluate_disp

**Reply by sitangshugk95** [reply]
1. What extra paramters does the contact problem solve for, compared to these methods? Is it the contact area, contact points etc.? 

2. If the answer to 1 is yes, I would need those paramters too (I worded it wrongly) - which code would be a good starting point for solving the contact problem for a rough surface taking a pressure distribution p(x.y) as input?

**Reply by sannant** [reply]
I do not understand anymore what you want to do, I am sorry. I think I already answered your question, or you should state much more clearly what you want to do. If the pressure distribution is prescribed there is no unknown to solve for and the displacements can be obtained by a simple convolution.
Or maybe should I point you to our paper on contact.engineering and the classical literature we mention in the README ? 
Maybe the textbook by Popov is helpful as well.

**Reply by sitangshugk95** [reply]
> Here are the methods you are looking for:
> 
> https://contactengineering.github.io/ContactMechanics/source/ContactMechanics.html?highlight=periodicfft#ContactMechanics.FFTElasticHalfSpace.PeriodicFFTElasticHalfSpace.evaluate_force
> 
> https://contactengineering.github.io/ContactMechanics/source/ContactMechanics.html?highlight=periodicfft#ContactMechanics.FFTElasticHalfSpace.PeriodicFFTElasticHalfSpace.evaluate_disp

I tried to validate this by doing the following:

1. Solve(sol) the contact problem for a given average pressure and get the force(sol.jac) and displacement arrays(sol.x).
2. Use the array of forces obtained from the solution as an input to the evaluate_disp(forces) function - if my understanding is correct, this should return the same displacement array that was previously obtained while solving the contact problem following this [function definition](https://github.com/ContactEngineering/ContactMechanics/blob/master/ContactMechanics/Optimization/ConstrainedConjugateGradients.py).

However, I am getting the following error:
![Screenshot from 2022-11-30 00-43-42](https://user-images.githubusercontent.com/108830413/204729363-e424afa8-d7f3-4311-98b9-12025fbb91ce.png)

I even tried applying zero force, and still received the same error:
![Screenshot from 2022-11-30 01-11-53](https://user-images.githubusercontent.com/108830413/204731222-922cf23d-b21c-4788-b3a5-f9e40d6d2aa3.png)

Am I using evaluate_disp() wrongly?
